Pynapple chapter (#14)
neuroinformatics-unit/course-ephys-osss: book/pynapple.qmd

# Neural analysis with Pynapple {.unnumbered}

This chapter is an introduction to [Pynapple](pynapple.org).  
Pynapple is a python package for analysing neural data.  
The package is actively maintained by the Flatiron Institute's [neuroRSE team](https://neurorse.flatironinstitute.org/). 

Pynapple can do a lot more than what is covered in this chapter, take a look at the [Pynapple documentation](https://pynapple.org/user_guide/03_core_methods.html) for more.   
The only thing you'll need to follow along is a working Python environment with `pynapple` installed:

```python

> pip install pynapple

Reviewer: At some point we can add a page on how to set up the environment for the entire course, and we can link to it from here and remove the `pip install pynapple`  code. Let's leave it for now and I'll make an issue from this comment.

Author: Maybe we can just keep it for ever but make the sentence say something like "If you haven't already...", linking to the page for sure. Mostly trying for the pynapple chapter to fit in the broader book, but perhaps also work on its own a tiny bit? 
But not too strongly opinionated on this, happy to make it entirely encapsulated in the book.

Reviewer: Sure let's keep it! That's a good point many people will likely also just work through a single chapter rather than the entire book, so makes sense to have some self-containedness and link to central resources for the bigger stuff

```

Now just open your favourite way of running Python, and follow along!

## Origins

- __2000:  TSToolbox, MATLAB__  
McNaughton lab: David Redish & Francesco Battaglia  

- __2016:  TSToolbox2, MATLAB__  
Adrien Peyrache & Luke Sjulson  

- __2018:  neuroseries,  Python__  
Francesco Battaglia  

- __2021:  pynapple, Python__  
Guillaume Viejo  

presently with the Flatiron Institute's neuroRSE team:

- __Guillaume Viejo__
- __Sarah Jo Venditto__
- Edoardo Balzani
- William Broderick

\+ __Wolf De Wulf__, a 2025 neuroRSE intern

![](https://githubcard.com/pynapple-org/pynapple.svg)

## Goal

> Pynapple was designed to lie in between pre- and postprocessing.  

Reviewer: would it be possible to use different terminology here as in previous chapters we have used the terms 'preprocessing' as ephys data preprocessing, and 'postprocessing' as spike-sorting output processing. Would it make sense to use Recording -> Processing -> Analysis here?

  <img src="images/rat.png" width="70%">
</div>

<div class="timeline-entry">
  <b>Preprocessing</b><br>
  segmenting (`calman`)<br>
  spike-sorting (`spikeinterface`)
</div>
<div class="timeline-entry logo">
  <img src="images/logo.png" width="100%">
</div>
<div class="timeline-entry">
  <b>Postprocessing</b><br>
  model fitting (`nemos`)<br>
  interpretation (`you`)
</div>
</div>
<div class="timeline-arrow"></div>
</div>
<style>
.timeline-horizontal {
  height: 20vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
}
.timeline-arrow {
  position: relative;
  width: 90%;
  height: 6px;
  background: currentColor;
  margin: 1em auto 0;
}
.timeline-arrow::after {
  content: "";
  position: absolute;
  right: -6px;
  top: -9px;
  border-top: 12px solid transparent;
  border-bottom: 12px solid transparent;
  border-left: 18px solid currentColor;
}
.timeline-entries-horizontal {
  display: grid;
  grid-template-columns: 1fr 1fr 0.5fr 1fr;
  align-items: center;
  width: 90%;
  margin: 0 auto 1em;
}
.timeline-entry {
  text-align: center;
  font-size: 1.1em;
}
.logo {
  margin: 0 1em;
}
</style>

## 1. Pynapple fundamentals

### Time series (`Ts`)

> We store a time series, for example a spike train, in a `Ts` object.

Reviewer: Out of interest does the `t` here have to be timestamps and can be extended to hold data with the `d` argument, or in theory could you store timeless data in `t` e.g. some voltage trace?

Author: Hmm no t should always be time! Voltage should go in the d of a `Tsd`.

```{python}

import pynapple as nap
ts = nap.Ts(t=[1,2,3,4,5])
ts
```
```{python}
#| echo: false

> import matplotlib.pyplot as plt

Reviewer: Here we could add 

```
#| code-fold: true
#| code-summary: "Show plotting code"
``` 

instead of `#| echo: false` and it will give the option to show the plotting code:

<img width="786" height="789" alt="Image" src="https://github.com/user-attachments/assets/c592167c-bdf8-41e6-9ede-1ee322b27f50" />

fig, ax = plt.subplots(constrained_layout=True, figsize=(5,3))
for timepoint in ts.times():
    ax.axvline(timepoint)
ax.yaxis.set_visible(False)
ax.spines['left'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
ax.set_xlabel("time (s)")
plt.show()
```

### Time series with data (`Tsd`)

> We store a time series with corresponding data values, for example a speed curve, in a `Tsd` object.

Reviewer: I think some attendees might not understand what a speed curve refers to, could this be expanded e.g. the speed of a mouse during escape from a threatening stimulus?

```{python}
import pynapple as nap
import numpy as np
tsd = nap.Tsd(t=[1,2,3,4,5], d=[1,1,1,2,1])
tsd
```
```{python}
#| echo: false

import matplotlib.pyplot as plt
fig, ax = plt.subplots(constrained_layout=True, figsize=(5,3))
ax.plot(tsd)
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.set_xlabel("time (s)")
ax.set_ylabel("data (a.u.)")
plt.show()
```

### Multiple time series
If we have multiple time series, they can be collected in a `TsGroup`.  

> This could be a collection of spiking units, for example.

Reviewer: ```suggestion
This could be a collection of spike times for multiple spike-sorted units, for example.
```

```{python}
import pynapple as nap
tsgroup = nap.TsGroup({
    1: nap.Ts([1,2,3,4,5]), 
    2: nap.Ts([1.5, 2.2, 2.9, 4.2])
})
tsgroup
```
```{python}
#| echo: false

import matplotlib.pyplot as plt
import numpy as np
cmap = plt.get_cmap("tab10")
fig, ax = plt.subplots(constrained_layout=True, figsize=(5,3))
for i in tsgroup:
    ax.vlines(tsgroup[i].times(), i, i+1, label=i, color=cmap(i-1))
ax.yaxis.set_visible(False)
ax.spines['left'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
ax.set_xlabel("time (s)")
plt.ylim(1,3)
plt.legend(frameon=False)
plt.show()
```

### Time series with multiple data columns (`TsdFrame`)

> If we have multiple time series with data with the same time points, they can be stored in a `TsdFrame`.  

Reviewer: ```suggestion
If we have a time series with multiple data associated with the same time points, they can be stored in a `TsdFrame`.  
```

> If we have multiple time series with data with the same time points, they can be stored in a `TsdFrame`.  

Reviewer: Here does each entry in `d` have to contain the same number of datapoints e.g. would the below be legal?

```
tsdframe = nap.TsdFrame(
    t=[1,2,3,4,5], 
    d=[[1,2],
       [1,2],
       [1,3],
       [2],
       [1]],
    columns=["A", "B"]
)
```

> This could be a collection of calcium traces, for example.

Reviewer: There might be attendees unfamiliar with calcium traces, could this be expanded to explain the significance (I guess different fluorescence channels, or traces from different cells)?

```{python}
import pynapple as nap
tsdframe = nap.TsdFrame(
    t=[1,2,3,4,5], 
    d=[[1,2],
       [1,2],
       [1,3],
       [2,1],
       [1,2]],
    columns=["A", "B"]
)
tsdframe
```
```{python}
#| echo: false
import matplotlib.pyplot as plt
import numpy as np
fig, ax = plt.subplots(constrained_layout=True, figsize=(5,3))
for i, col in enumerate(tsdframe.columns):
    ax.plot(tsdframe[:, i], label=col)
ax.set_ylabel("data (a.u.)")
ax.set_xlabel("time (s)")
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)
plt.legend(frameon=False)
plt.show()
```

### Intervals (`IntervalSet`)

> We store time intervals as start and end pairs in an `IntervalSet`.

Reviewer: Maybe a brief motivating example here e.g. Often we want to store events that occur across a period rather than at single time points, for example a behavioural stimulus that lasts multiple seconds. To represent this, we can....

```{python}
import pynapple as nap
epochs = nap.IntervalSet(
    start=[1, 3, 7],
    end=[2, 5, 9]
)
epochs
```
```{python}
#| echo: false

import matplotlib.pyplot as plt
fig, ax = plt.subplots(constrained_layout=True, figsize=(5,3))
for start, stop in epochs.values:
    ax.axvspan(start, stop)
ax.set_xlabel("time (s)")
ax.spines['left'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
plt.show()
```

## 2. Interactions

> Many Pynapple objects interact with each other.

Reviewer: Maybe here a quick motivating statement e.g. A key aim/purpose of Pynapple is to integrate electrophysiological and behavioural data for processing and analysis. Therefore, many...

### Time supports (`time_support`)
Every time series object has a time support, an `IntervalSet` that gives the epochs
over which the time series is defined.
```{python}
import pynapple as nap
ts = nap.Ts(t=[1,2,3,4,5])
ts.time_support
```
```{python}
#| echo: false
import matplotlib.pyplot as plt
fig, ax = plt.subplots(constrained_layout=True, figsize=(5,3))
for timepoint in ts.times():
    ax.axvline(timepoint)
for start, stop in ts.time_support.values:
    ax.axvspan(start, stop, alpha=0.2)
ax.yaxis.set_visible(False)
ax.spines['left'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
ax.set_xlabel("time (s)")
plt.show()
```

### Restricting (`restrict`)

> Any time series object can be restricted to certain `IntervalSet`, thereby updating its time support.

Reviewer: maybe an example? e.g. This could be used, for example, to keep only the spike times which fall within the presentation of a behavioural stimulus

```{python}
import pynapple as nap
ts = nap.Ts(t=[1,2,3,4,5])
restriction = nap.IntervalSet(start=[1.3, 3.9], end=[3.5, 4.2])
restricted = ts.restrict(restriction)
restricted
```

> ```{python}

Reviewer: cool!

I wonder in the image if it makes sense to set the xlim to 1-5 so it is easy to compare with the previous figure what has been removed?

#| echo: false
import matplotlib.pyplot as plt
fig, ax = plt.subplots(constrained_layout=True, figsize=(5,3))
for timepoint in restricted.times():
    ax.axvline(timepoint)
for start, stop in restricted.time_support.values:
    ax.axvspan(start, stop, alpha=0.2)
ax.yaxis.set_visible(False)
ax.spines['left'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
ax.set_xlabel("time (s)")
plt.show()
```

### Taking values (`value_from`)
Given one time series, you can `value_from` to take the value corresponding 

> to the closest time point in another time series with data.

Reviewer: could this be used to sample the position of an animal at particular spike times?

Author: Yes, that is actually one of the most typical use-cases for this function.
Added a sentence with that as an example!

    t=[1,2,3,4,5], 
    d=[[1,2],
       [1,2],
       [1,3],
       [2,1],
       [1,2]],
    columns=["A", "B"]
)
np.mean(tsdframe, axis=1)
```
```{python}

import pynapple as nap
import numpy as np
tsdframe = nap.TsdFrame(
    t=[1,2,3,4,5], 
    d=[[1,2],
       [1,2],
       [1,3],
       [2,1],
       [1,2]],
    columns=["A", "B"]
)
np.diff(tsdframe, axis=1)
```

### Slicing
Pynapple objects can be sliced like Numpy arrays.
```{python}
import pynapple as nap
import numpy as np
tsdframe = nap.TsdFrame(
    t=[1,2,3,4,5], 
    d=[[1,2],
       [1,2],
       [1,3],
       [2,1],
       [1,2]],
    columns=["A", "B"]
)
tsdframe[:, 1]
```
```{python}

import pynapple as nap
import numpy as np
tsdframe = nap.TsdFrame(
    t=[1,2,3,4,5], 
    d=[[1,2],
       [1,2],
       [1,3],
       [2,1],
       [1,2]],
    columns=["A", "B"]
)
tsdframe[2:4]
```

## 4. Core functions
Pynapple has a core set of functions designed to help with neuroscientific analyses.
The following are a subset of them, for an exhaustive list, see the [Pynapple documentation](https://pynapple.org/user_guide/03_core_methods.html).

### Binning
The time points of any time series object can be counted.

> ```{python}

Reviewer: This might be getting towards a technical level that some attendees may not be familiar with and have trouble visualising, maybe to ease understanding we can expand some variable names / add comments here e.g.

```
import pynapple as nap
import numpy as np
time_step_s= 0.001
times = np.arange(0, 4, time_step_s)

# Create a firing pattern that increases and decreases as a sine wave
rate = 40 * (1 + np.sin(2*np.pi*times)) / 2
spikes = nap.Ts(times[np.random.rand(len(times)) < rate * time_step_s])

# Bin the number of spikes
counts = spikes.count(bin_size=0.1)

```

maybe we could plot the modulating sinusoid just to help people understand this cool way of generating sine-modulated spike times? although that been said don't want to detract from the main point of this.

Author: absolutely, I've added an axes for the rate!

import pynapple as nap
import numpy as np
dt = 0.001
times = np.arange(0, 4, dt)
rate = 40 * (1 + np.sin(2*np.pi*times)) / 2
spikes = nap.Ts(times[np.random.rand(len(times)) < rate * dt])
counts = spikes.count(bin_size=0.1)
counts
```
```{python}
#| echo: false

import matplotlib.pyplot as plt
fig, (ax_spikes, ax_count) = plt.subplots(2, 1, constrained_layout=True, figsize=(5,3), sharex=True)
ax_spikes.vlines(spikes.times(), 0, 1)
ax_spikes.set_ylabel("spikes")
ax_spikes.set_yticks([])
ax_spikes.spines['left'].set_visible(False)
ax_spikes.spines['right'].set_visible(False)
ax_spikes.spines['top'].set_visible(False)
ax_count.plot(counts)
ax_count.set_ylabel("count")
ax_count.set_xlabel("time (s)")
ax_count.spines['right'].set_visible(False)
ax_count.spines['top'].set_visible(False)
plt.show()
```

### Averaging

> The values of any time series with data can be averaged across bins.

Reviewer: I'm not sure if it makes sense to, but would it work to carry forward the example from the previous 'Binning' section and thread it through with an example use case so its a bit of a story e.g. we have spiking when a oscillating visual stimulus is displayed and we want to i) bin the spikes ii) compute the average across multiple bins to reduce data size  iii) find periods when the firing exceeds some threshold

Author: Hmmm I'd enjoy that, but it is a bit atypical, as in, you'd use a bigger bin size if you wanted to reduce data size. The averaging is more of a parallel for continuous data. I'll try and word it that way!

Reviewer: sure makes sense!

```{python}
import pynapple as nap
import numpy as np

> times = np.arange(0, 10, 0.01)

Reviewer: could add `time_step_s` variable here just for consistency with other code blocks

fig, ax = plt.subplots(constrained_layout=True, figsize=(5,3))
plt.plot(tsd, label="original")
plt.plot(averaged, label="averaged")
ax.set_ylabel("data (a.u.)")
ax.set_xlabel("time (s)")
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
fig.legend(frameon=False, bbox_to_anchor=(1.0, 1.1))
plt.show()
```

### Thresholding
Time series with data can be thresholded, yielding new time series objects with updated time supports.
```{python}
import pynapple as nap
import numpy as np
times = np.arange(0, 10, 0.01)
signal = np.sin(2 * np.pi * times) 
signal += 0.2 * np.random.randn(len(times))
tsd = nap.Tsd(t=times, d=signal)
above = tsd.threshold(0.5, method="above")
below = tsd.threshold(-.5, method="below")
above.time_support
```
```{python}
#| echo: false

import matplotlib.pyplot as plt
fig, ax = plt.subplots(constrained_layout=True, figsize=(5,3))
ax.plot(tsd, label="original")
ax.plot(above, label="above", marker=".", linestyle="none")
ax.plot(below, label="below", marker=".", linestyle="none")
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
fig.legend(frameon=False, bbox_to_anchor=(1.0, 1.2))
plt.show()
```

### Smoothing

> Time series with data can be (Gaussian) smoothed.

Reviewer: Maybe a very brief expansion here on what the smoothing operation does and its usefulness in the data processing context

```{python}
import pynapple as nap
import numpy as np
times = np.arange(0, 10, 0.01)
signal = np.sin(2 * np.pi * times) 
signal += 0.2 * np.random.randn(len(times))
tsd = nap.Tsd(t=times, d=signal)
smoothed = tsd.smooth(std=0.1)
smoothed
```
```{python}
#| echo: false

import matplotlib.pyplot as plt
fig, ax = plt.subplots(constrained_layout=True, figsize=(5,3))
ax.plot(tsd, label="original")
ax.plot(smoothed, label="smoothed")
ax.set_ylabel("data (a.u.)")
ax.set_xlabel("time (s)")
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
fig.legend(frameon=False, bbox_to_anchor=(1.0, 1.1))
plt.show()
```

## 5. Advanced Functions
Beyond fundamental processing functions, Pynapple has loads of advanced functionality for neuroscientific analysis.

### Autocorrelograms

> Given one or a group of time series, you can compute autocorrelograms.

Reviewer: For me to make an issue from this comment, here we could link to a section we will have in a previous chapter (e.g. maybe postprocessing) that introduces the autocorrelogram, then mention here we are similarly computing unit autocorrelograms.

Author: I'm adding big red notes where this needs to happen!

Reviewer: brilliant cheers

import numpy as np
dt = 0.01
times = np.arange(0, 4, dt)
rate = 40 * (1 + np.sin(2*np.pi*times)) / 2
units = nap.TsGroup({
    1: nap.Ts(times[np.random.rand(len(times)) < rate * dt]),
    2: nap.Ts(times[np.random.rand(len(times)) < rate*2 * dt]),
})
autocorrelogram = nap.compute_autocorrelogram(
    units, binsize=0.1, windowsize=1.0, norm=False
)
autocorrelogram
```
```{python}
#| echo: false

import matplotlib.pyplot as plt
ax=autocorrelogram.plot()
plt.ylabel("rate (Hz)")
plt.xlabel("time (s)")
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
plt.legend(frameon=False)
plt.show()
```
### Crosscorrelograms
Given one or two groups of time series, you can compute crosscorrelograms.
```{python}
crosscorrelogram = nap.compute_crosscorrelogram(
    units, binsize=0.1, windowsize=1.0, norm=False
)
crosscorrelogram
```
```{python}
#| echo: false
import matplotlib.pyplot as plt
ax=crosscorrelogram.plot()
plt.ylabel("rate (Hz)")
plt.xlabel("time (s)")
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
plt.legend(frameon=False)
plt.show()
```

### Signal processing
Pynapple wraps a lot of scipy functions for signal processing.

> Let's simulate a noisy multi-frequency signal:

Reviewer: ```suggestion
Let's simulate a noisy multi-frequency signal (the combination of multiple sine waves oscillating at different frequencies):
```

```{python}
import pynapple as nap
import numpy as np
fs = 1000
t = np.linspace(0, 2, fs * 2)
f2 = np.cos(t*2*np.pi*2)
f10 = np.cos(t*2*np.pi*10)
f50 = np.cos(t*2*np.pi*50)
tsd = nap.Tsd(t=t,d=f2+f10+f50 + np.random.normal(0, 0.5, len(t)))
tsd
```
```{python}

> #| echo: false

Reviewer: I think some attendees might not be super familiar with this, does it make sense to include the constituent sine waves in the plots so that it is clear what this multi-frequency signal is made up of? this might also make it clear how some operations (e.g. filtering) 'reveal' its parts

import matplotlib.pyplot as plt
fig, ax = plt.subplots(constrained_layout=True, figsize=(5,3))
ax.plot(tsd)
ax.set_ylabel("data (a.u.)")
ax.set_xlabel("time (s)")
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
plt.show()
```

#### Bandpass filter

> We can bandpass filter to exract certain frequency bands.

Reviewer: Maybe a quick expansion just to spell it out to those unfamiliar e.g. for example, if we apply a bandpass filter with a 8Hz high-pass and 12Hz low-pass cutoff, we will keep only the part of the signal in the 8-12hz range

```{python}
filtered = nap.apply_bandpass_filter(
    data=tsd, 
    cutoff=(8, 12), 
    fs=fs, 
    mode='butter'
)
filtered
```
```{python}
#| echo: false

import matplotlib.pyplot as plt
fig, ax = plt.subplots(constrained_layout=True, figsize=(5,3))
ax.plot(filtered)
ax.set_ylabel("data (a.u.)")
ax.set_xlabel("time (s)")
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
plt.show()
```

#### Power-spectral density

> We can compute the power-spectral density to find frequencies of interest.

Reviewer: Same here although ofc we cannot do a comprehensive explanation of the PSD. However maybe something on how it shows the power of a signal at frequencies up to 0.5 the sampling rate, with a link to a more thorough explanation?

```{python}
psd = nap.compute_power_spectral_density(
    tsd, 
    fs=fs
)
psd
```
```{python}
#| echo: false
import matplotlib.pyplot as plt
fig, ax = plt.subplots(constrained_layout=True, figsize=(5,3))
ax.plot(psd)
ax.set_ylabel("power")
ax.set_xlabel("frequency (Hz)")
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
plt.show()
```

### Perievent

> A typical neuroscientific analysis consists of aligning activity to a set of stimuli or events.

Reviewer: Maybe we could motivate here with a real world example e.g. for example this may be spikes around cue presentation for a food-associated cue

Let's start by simulating a spiking unit that has clear stimulus-locked activity:
```{python}
import pynapple as nap
import numpy as np
stimuli = nap.Ts(t=np.arange(0, 1000, 1), time_units="s")
baseline = np.random.uniform(0, 1000, 500)
burst = np.concatenate([
        np.random.normal(
            st + 0.1, 0.05, 3
        ) 
        for st in stimuli.times()
    ])
ts = nap.Ts(t=np.sort(np.concatenate([baseline, burst])))
ts
```
```{python}
#| echo: false
import matplotlib.pyplot as plt
segment = nap.IntervalSet(100, 103.9)
fig, ax = plt.subplots(1, 1, constrained_layout=True)
ax.vlines(ts.restrict(segment).times(), 0.04, 0.10, label="spikes")
ax.vlines(
    stimuli.restrict(segment).times(),
    0.0,
    0.14,
    color="gray",
    linestyle="--",
    label="stimulus",
)
ax.yaxis.set_visible(False)
ax.spines["left"].set_visible(False)
ax.spines['right'].set_visible(False)
ax.spines['top'].set_visible(False)
ax.set_xlabel("time (s)")
plt.show()
```

> Now, we can easily compute a peri-event time histogram (PETH) as follows:

Reviewer: Could we plot here too? and then plot again after binning to see the difference. I think here for those unfamiliar the exact construction of the PETH could be explained e.g. we are stacking all trials on top of each other

```{python}
peth = nap.compute_perievent(
    data=ts, 
    events=stimuli, 
    window=(-0.1, 0.4))
peth
```

We can also bin the spikes:
```{python}
bin_size = 0.01
peth_counts = peth.count(bin_size)
peth_counts
```

Visualising everything together:
```{python}
#| echo: false
import matplotlib.pyplot as plt
fig, (ax_spikes, ax_mean) = plt.subplots(
    2, 1, sharex=True, height_ratios=[1, 0.3], figsize=(6, 6), gridspec_kw={"hspace": 0.3}
)
mean = np.mean(peth_counts / bin_size, axis=1)
ax_mean.plot(mean)
ax_mean.set_ylabel("spikes/s")
ax_mean.axvline(0.0, color="gray", linestyle="--")
ax_mean.spines['right'].set_visible(False)
ax_mean.spines['top'].set_visible(False)
ax_spikes.plot(peth.to_tsd(), "|", markersize=5)
ax_mean.set_xlabel("time from event (s)")
ax_spikes.set_ylabel("event")
ax_spikes.axvline(0.0, color="gray", linestyle="--")
ax_spikes.spines['right'].set_visible(False)
ax_spikes.spines['top'].set_visible(False)
plt.show()
```

### Tuning curves

> Let's start by simulating some spiking units that are clearly modulated by a circular feature:

Reviewer: This example is really nice, I think for some attendees the head direction set up and / or tuning curves may be quite new ideas. Maybe this section could be expanded slightly to briefly set up the context of head direction cells and maybe include a graphic like below?

<img width="2015" height="1892" alt="Image" src="https://github.com/user-attachments/assets/f7d5c8b3-3dae-40ca-bd32-c30ff8584303" />

Although the mouses heads have a bit of a silence of the lamb vibe 😂 

Maybe with the code stimulus generation as well the `tmp` and `tc` names expanded with some explanatory comments, and maybe a plot to show the unit tuning curves like

<img width="1081" height="994" alt="Image" src="https://github.com/user-attachments/assets/9924b9a7-3faa-4f5d-9451-350c694ada5c" />

(or something less extensive, maybe the first subplot)

Author: Hmm I see what you are saying, but that is already assuming it is head direction, whereas here I want to stick to a circular variable. The notebook uses head direction. Perhaps I should switch to 1D position here? That might be easier to simulate as well

Reviewer: Yes I was thinking its nice to have this a bit more abstract and then make it concrete with the notebooks using the head-direction realisation. However I think some attendees will have trouble understanding the concept of tuning to a circular variable without an example application. If 1D position works well for this and is okay with you that would be perfect!

Author: Yep fair, I'm going to think about it a bit more, I'd actually prefer sticking to the circular variable. But I'll see how to make the data generation more understandable

```{python}
import pandas as pd
import pynapple as nap
import numpy as np
from scipy.ndimage import gaussian_filter1d

N = 6 # Number of neurons
bins = np.linspace(0, 2*np.pi, 61)
x = np.linspace(-np.pi, np.pi, len(bins)-1)
tmp = np.roll(np.exp(-(1.5*x)**2), (len(bins)-1)//2)
tc = np.array([np.roll(tmp, i*(len(bins)-1)//N) for i in range(N)]).T

tc_1d = pd.DataFrame(index=bins[0:-1], data=tc)

# Feature
T = 10000
dt = 0.01
timestep = np.arange(0, T)*dt
feature = nap.Tsd(
    t=timestep,
    d=gaussian_filter1d(np.cumsum(np.random.randn(T)*0.5), 20)%(2*np.pi)
    )
index = np.digitize(feature, bins)-1

# Spiking activity
count = np.random.poisson(tc[index])>0
tsgroup = nap.TsGroup({i:nap.Ts(timestep[count[:,i]]) for i in range(N)})
epochs = nap.IntervalSet(0, 10)
tsgroup
```

We can use Pynapple to easily compute tuning curves with respect to the circular feature.  
The output is an [`xarray.DataArray`](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.html):
```{python}
tuning_curves = nap.compute_tuning_curves(
    data=tsgroup,
    features=feature,
    bins=120, 
    range=(0, 2*np.pi),
    feature_names=["feature"]
    )
tuning_curves
```

> These can be easily visualised:

Reviewer: Maybe a quick explanation of what the plot is showing e.g. we can see how the different neurons respond when the animals head is at different angles

```{python}
#| output: false
import matplotlib.pyplot as plt
tuning_curves.plot.line(x="feature", add_legend=False)
plt.ylabel("firing rate [Hz]")
```
```{python}
#| echo: false
tuning_curves.plot.line(x="feature", add_legend=False)
plt.ylabel("firing rate [Hz]")
plt.xticks([0, 2*np.pi], ["0", "2π"])
plt.show()
```

> ### Decoding

Reviewer: Here I think we for some people can assume very little previous experience with neural decoding. Could we expand this a little just to give a quick overview of Bayesian decoding and spell out what is happening here

```{python}
decoded, proba_feature = nap.decode_bayes(
    tuning_curves=tuning_curves,
    data=tsgroup,
    epochs=epochs,
    sliding_window_size=4,
    bin_size=0.02,
)
decoded
```
```{python}
#| echo: false
import matplotlib.pyplot as plt
fig, (ax1, ax2) = plt.subplots(figsize=(6, 5), nrows=2, ncols=1, sharex=True)
ax1.plot(
    feature.times(),
    feature.values,
    label="true",
)
ax1.scatter(
    decoded.times(),
    decoded.values,
    label="decoded",
    c="orange",
)
ax1.spines['right'].set_visible(False)
ax1.spines['top'].set_visible(False)
fig.legend(frameon=False, bbox_to_anchor=(0.9, 1.0))
ax1.set_ylabel("feature")
ax1.set_yticks([0, 2*np.pi], ["0", "2π"])
proba_feature = (proba_feature.values - np.min(proba_feature.values, axis=1, keepdims=True)) / np.ptp(proba_feature.values, axis=1,keepdims=True)
im = ax2.imshow(proba_feature.T, aspect="auto", origin="lower", cmap="viridis", extent=(0, 10.0, 0, 2*np.pi))
cbar_ax = fig.add_axes([0.93, 0.1, 0.015, 0.36])
fig.colorbar(im, cax=cbar_ax, label="probability")
ax2.set_xlabel("time (s)")
ax2.set_ylabel("feature")
ax2.set_yticks([0, 2*np.pi], ["0", "2π"])
plt.show()
```

## 6. Integration with other packages

> ### Pynaviz

Reviewer: cool!

[Pynaviz](https://github.com/pynapple-org/pynaviz) provides interactive, high-performance visualisations designed to work seamlessly with Pynapple time series and video data.   
It allows synchronised exploration of neural signals and behavioral recordings.   
It is build on top of `pygfx`, a modern GPU-based rendering engine.

To install it:
```python
pip install pynaviz[qt]
```

### NeMoS
![](https://raw.githubusercontent.com/flatironinstitute/nemos/main/docs/assets/glm_features_scheme.svg)

[NeMoS](https://nemos.readthedocs.io/en/latest/index.html) (Neural ModelS) is a statistical modeling framework optimized for systems neuroscience and powered by JAX.  
It streamlines the process of defining and selecting models, through a collection of easy-to-use methods for feature design.

The core of `nemos` includes GPU-accelerated, well-tested implementations of standard statistical models, currently focusing on the Generalized Linear Model (GLM).  
Check out [this page](https://nemos.readthedocs.io/en/latest/tutorials/README.html) for many examples of neural modelling using `nemos` and `pynapple`.

To install it:
```python
pip install nemos
```
### SpikeInterface

> [SpikeInterface](https://spikeinterface.readthedocs.io/en/latest/) is a Python library for spike sorting and electrophysiological data analysis.

Reviewer: I think we can remove this line as SpikeInterface will be introduced earlier in the book

SpikeInterface can export the output of spike sorting to `pynapple`, allowing you to seamlessly integrate spike sorting results into your pynapple-based analysis pipeline:
```python
import spikeinterface as si
from spikeinterface.exporters import to_pynapple_tsgroup

analyzer = si.load_sorting_analyzer("path/to/analyzer")
tsgroup = to_pynapple_tsgroup(
    sorting_analyzer=analyzer,
    attach_unit_metadata=True,
)
```

## 7. Data
Pynapple provides many ways to load your data.

### NWB

> The main input format Pynapple can read is [Neurodata Without Borders (NWB)](https://nwb.org/).  

Reviewer: It could be worth mentioning here that NEO is the same library SpikeInterface uses under the hood for a lot of data formats

Pynapple will parse all data in the file that can be represented as a Pynapple object and present a dictionary-like interface for selecting them.
```{python}
#| echo: false
import os
import requests
nwb_path = 'A2929-200711.nwb'

if nwb_path not in os.listdir("."):
  r = requests.get(f"https://osf.io/fqht6/download", stream=True)
  block_size = 1024*1024
  with open(nwb_path, 'wb') as f:
    for data in r.iter_content(block_size):
        f.write(data)
```
```{python}
import pynapple as nap
nwb_path = 'A2929-200711.nwb'
data = nap.load_file(nwb_path)
data
```

### Raw data
Raw data can be loaded with `pynapple` through the [NEO](https://neo.readthedocs.io/en/latest/) library.   
Many formats are supported, [see here](https://neo.readthedocs.io/en/stable/rawiolist.html).
```python
import pynapple as nap
data = nap.EphysReader("path/to/raw")
```

### Saving as `.npz`

> Pynapple objects can be saved (and loaded) as `.npz` files.

Reviewer: ```suggestion
Pynapple objects can be saved (and loaded) as numpy `.npz` files.
```

(just in case people are not familiar)

Here is useful function to keep handy that does this for you:
```{python}
import pynapple as nap
from pynwb import NWBHDF5IO
from dandi.dandiapi import DandiAPIClient
import fsspec
from fsspec.implementations.cached import CachingFileSystem
import h5py

def stream_dandiset(dandiset_id, filepath):
    with DandiAPIClient() as client:
        asset = client.get_dandiset(dandiset_id, "draft").get_asset_by_path(filepath)
        s3_url = asset.get_content_url(follow_redirects=1, strip_query=True)

    # first, create a virtual filesystem based on the http protocol
    fs = fsspec.filesystem("http")

    # create a cache to save downloaded data to disk (optional)
    fs = CachingFileSystem(
        fs=fs,
        cache_storage="nwb-cache",  # Local folder for the cache
    )

    # next, open the file
    file = h5py.File(fs.open(s3_url, "rb"))
    io = NWBHDF5IO(file=file, load_namespaces=True)

    return nap.NWBFile(io.read())
```

## Your turn
Now it is your turn!

Start by streaming a single session from [this dataset](https://dandiarchive.org/dandiset/000939?search=000939&pos=1) on the DANDI archive, it contains recordings from head direction cells in the postsubiculum.

> Use the streaming function we provided in the data section!

Reviewer: Reminder for me to make an issue from this comment, link to the data page from here
